Sources/SecureXPC/Server/XPCClientRequirement.swift

```diff
@@ -8,88 +8,364 @@
+        do {
+            try expandSandboxIfNecessary(message: message)
+        } catch {
+            return false
+        }
+        
+        guard let code = SecCodeCreateWithXPCConnection(connection, andMessage: message) else {
+            return false
+        }
+        
+        return self.requirements.contains { SecCodeCheckValidity(code, SecCSFlags(), $0) == errSecSuccess }
+    }
+    
+    func isEqual(to acceptor: MessageAcceptor) -> Bool {
+        guard let acceptor = acceptor as? SecRequirementsMessageAcceptor else {
+            return false
+        }
+        
+        // Transform the requirements into Data form so that they can be compared
+        let requirementTransform = { (requirement: SecRequirement) -> Data? in
+            var data: CFData?
+            if SecRequirementCopyData(requirement, SecCSFlags(), &data) == errSecSuccess, let data = data as Data? {
+                return data
+            } else {
+                return nil
+            }
+        }
+        
+        // Turn into sets so they can be compared without taking into account the order of requirements
+        let requirementsData = Set<Data>(self.requirements.compactMap(requirementTransform))
+        let otherRequirementsData = Set<Data>(acceptor.requirements.compactMap(requirementTransform))
+        
+        return requirementsData == otherRequirementsData
+    }
+}
+
+/// Accepts messages that originates from a directory containing this server.
+fileprivate struct ParentBundleMessageAcceptor: MessageAcceptor {
+    /// Requests must come from a process that either has the same URL as this or is a subdirectory of it.
+    private let parentBundleURL: URL
+    
+    init(parentBundleURL: URL) {
+        self.parentBundleURL = parentBundleURL
+    }
+    
+    func acceptMessage(connection: xpc_connection_t, message: xpc_object_t) -> Bool {
+        do {
+            try expandSandboxIfNecessary(message: message)
+        } catch {
+            return false
+        }
+        
+        guard let clientCode = SecCodeCreateWithXPCConnection(connection, andMessage: message),
+              let clientPath = SecCodeCopyPath(clientCode) else {
+            return false
+        }
+        
+        // If this is true there's no possibility of the client being equal to or a subdirectory of the parent bundle.
+        // And importantly, this prevents indexing out of bounds when iterate through to check equality/containment.
+        if clientPath.pathComponents.count < parentBundleURL.pathComponents.count {
+            return false
+        }
+        
+        // Validate the each component in the client path is present in the parent bundle path in the same order
+        for i in 0..<parentBundleURL.pathComponents.count {
+            if parentBundleURL.pathComponents[i] != clientPath.pathComponents[i] {
+                return false
+            }
+        }
+        
+        return true
```

Comment:
You can replace this manual indexing with `zip`, and this condition checking with `allSatisfy(_:)` (in other cases where you conditionally `return true` else `return false`, you can use `contains(where:)`)

```suggestion
        return zip(parentBundleURL.pathComponents, clientPath.pathComponents).allSatisfy(==)
```

Reply:
Interesting, I didn't realize `zip` could handle inputs of two different lengths and that it would handle the shorter one. That's quite convenient, although I do think in some ways it makes the code a little harder to reason about due to need to remember that aspect of it. Nonetheless, will switch over to using it.

Reply:
Indeed. I think that's relative well known, so it's safe to rely on.

The less obvious part is what to do if you want to have it pair up to the length of the longer sequence (e.g. pairing the elements of the longer sequence with `nil`). There's no built in solution for that.